File: weile_ddz/vision.py
"""
YOLO / 截图识别预留接口。

本版不跑模型、不抓屏、不控制外部程序。
未来若接入：用户主动上传图片 → 识别牌名 → 回填 CardState，并必须人工修正。
微乐是游戏渲染牌面，公开实体扑克权重通常不适用，需要自己标注微调。
"""

from __future__ import annotations

import re
from typing import Any

from weile_ddz.cards import CARD_BY_ID
from weile_ddz.state import Bucket, CardState

SCHEMA_VERSION = 1

LOCATION_TO_BUCKET = {
    "hand": Bucket.IN_HAND,
    "in_hand": Bucket.IN_HAND,
    "seen": Bucket.SEEN,
    "played": Bucket.SEEN,
    "unknown": Bucket.UNKNOWN,
}


class VisionNotConnected(RuntimeError):
    pass


def resolve_label(label: str) -> str | None:
    raw = (label or "").strip()
    if raw in CARD_BY_ID:
        return raw
    aliases = {
        "小王": "SJ",
        "大王": "BJ",
        "joker_small": "SJ",
        "joker_big": "BJ",
        "JOX": "SJ",
        "JOY": "BJ",
        "SJ": "SJ",
        "BJ": "BJ",
    }
    if raw in aliases:
        return aliases[raw]
    compact = raw.replace(" ", "")
    if compact in CARD_BY_ID:
        return compact
    match = re.fullmatch(r"([SHCD])_?([0-9]{1,2}|A|J|Q|K)", compact, flags=re.I)
    if match:
        suit = match.group(1).upper()
        rank = match.group(2).upper()
        if rank == "1":
            return None
        if rank == "11":
            rank = "J"
        if rank == "12":
            rank = "Q"
        if rank == "13":
            rank = "K"
        if rank == "14":
            rank = "A"
        if rank == "15":
            rank = "2"
        card_id = f"{suit}_{rank}"
        if card_id in CARD_BY_ID:
            return card_id
    return None


def ingest_detections(state: CardState, payload: dict[str, Any], threshold: float = 0.5) -> dict[str, Any]:
    detections = payload.get("detections") or payload.get("cards") or []
    applied = []
    skipped = []
    for item in detections:
        card_id = resolve_label(item.get("label") or item.get("class") or "")
        conf = float(item.get("confidence", item.get("score", 1)))
        loc = str(item.get("location", "unknown")).lower()
        bucket = LOCATION_TO_BUCKET.get(loc)
        if not card_id:
            skipped.append({**item, "reason": "无法映射牌面"})
            continue
        if bucket is None:
            skipped.append({**item, "reason": f"未知位置 {loc}"})
            continue
        if conf < threshold:
            skipped.append({**item, "reason": "置信度过低"})
            continue
        state.set_card(card_id, bucket, record_undo=False)
        applied.append({"id": card_id, "bucket": bucket.value, "confidence": conf})
    return {"applied": applied, "skipped": skipped, "message": f"已合并 {len(applied)} 张，跳过 {len(skipped)} 条"}


def recognize_image(_image_bytes: bytes) -> dict[str, Any]:
    raise VisionNotConnected(
        "YOLO 尚未接入。请继续手动点选；若以后上传截图识别，也必须人工核对结果。"
    )

File: js/vision.js
/**
 * YOLO / 图像识别预留接口。
 *
 * 本版只接受「已经识别好的结构化结果」，不读游戏内存、
 * 不挂钩微乐进程、不自动截屏、不自动点击。
 *
 * 未来接入方式：
 * 1. 独立 YOLO 模块输出 JSON（见 VISION_SCHEMA）
 * 2. 调用 window.__WEILE_VISION_BRIDGE__.applyDetections(payload)
 * 3. 或调用 ingestVisionPayload(state, payload)
 */
import { CARD_BY_ID, DECK } from "./cards.js";
import { BUCKET } from "./tracker.js";

export const VISION_SCHEMA_VERSION = 1;

export const VISION_SCHEMA = {
  schemaVersion: VISION_SCHEMA_VERSION,
  source: "yolo",
  threshold: 0.5,
  detections: [
    {
      label: "S3",
      confidence: 0.92,
      location: "hand",
      bbox: [0, 0, 0, 0],
    },
  ],
};

const LOCATION_TO_BUCKET = {
  hand: BUCKET.HAND,
  mine: BUCKET.HAND,
  my_hand: BUCKET.HAND,
  played: BUCKET.PLAYED,
  discarded: BUCKET.PLAYED,
  table: BUCKET.PLAYED,
  unknown: BUCKET.UNKNOWN,
  opponent: BUCKET.UNKNOWN,
  bottom: BUCKET.UNKNOWN,
  landlord_bottom: BUCKET.UNKNOWN,
};

function buildAliasMap() {
  const map = {};
  for (const card of DECK) {
    map[card.id] = card.id;
    map[card.id.toLowerCase()] = card.id;
    map[card.name] = card.id;
    if (!card.isJoker) {
      map[`${card.suit}${card.label}`] = card.id;
      map[`${card.suit.toLowerCase()}${card.label}`] = card.id;
      map[`${card.suit.toLowerCase()}${card.rank}`] = card.id;
      map[`${card.symbol}${card.label}`] = card.id;
      map[`${card.label}${card.suit}`] = card.id;
    }
  }
  const extras = {
    joker_small: "JOX",
    joker_big: "JOY",
    small_joker: "JOX",
    big_joker: "JOY",
    joker_s: "JOX",
    joker_b: "JOY",
    jox: "JOX",
    joy: "JOY",
    小王: "JOX",
    大王: "JOY",
    black_joker: "JOX",
    red_joker: "JOY",
    "joker-small": "JOX",
    "joker-big": "JOY",
  };
  Object.assign(map, extras);
  return map;
}

export const YOLO_LABEL_ALIASES = buildAliasMap();

export function resolveVisionLabel(label) {
  if (!label) return null;
  const raw = String(label).trim();
  if (CARD_BY_ID[raw]) return raw;
  const key = raw.replace(/\s+/g, "");
  return YOLO_LABEL_ALIASES[key] || YOLO_LABEL_ALIASES[key.toLowerCase()] || null;
}

export function parseVisionPayload(payload, threshold = 0.5) {
  if (!payload || typeof payload !== "object") {
    throw new Error("识别结果必须是对象");
  }
  const detections = payload.detections || payload.cards || [];
  if (!Array.isArray(detections)) {
    throw new Error("识别列表必须是数组");
  }
  const cut = payload.threshold ?? threshold;
  const accepted = [];
  const skipped = [];
  for (const item of detections) {
    const id = resolveVisionLabel(item.label || item.class || item.name);
    const confidence = item.confidence ?? item.score ?? 1;
    const location = String(item.location || item.bucket || "unknown").toLowerCase();
    const bucket = LOCATION_TO_BUCKET[location];
    if (!id) {
      skipped.push({ ...item, reason: "无法映射牌面" });
      continue;
    }
    if (!bucket) {
      skipped.push({ ...item, reason: `未知位置 ${location}` });
      continue;
    }
    if (confidence < cut) {
      skipped.push({ ...item, reason: `置信度 ${confidence} 低于阈值` });
      continue;
    }
    accepted.push({
      id,
      bucket,
      confidence,
      location,
      bbox: item.bbox || null,
    });
  }
  return {
    schemaVersion: payload.schemaVersion ?? VISION_SCHEMA_VERSION,
    source: payload.source || "yolo",
    accepted,
    skipped,
  };
}

/**
 * 把识别结果合并进记牌状态。
 * 同一张牌多次出现时：手牌优先于已出，已出优先于未知。
 */
export function ingestVisionPayload(state, payload, options = {}) {
  const parsed = parseVisionPayload(payload, options.threshold ?? 0.5);
  const priority = { [BUCKET.HAND]: 3, [BUCKET.PLAYED]: 2, [BUCKET.UNKNOWN]: 1 };
  const best = new Map();
  for (const hit of parsed.accepted) {
    const prev = best.get(hit.id);
    if (!prev || priority[hit.bucket] > priority[prev.bucket] || hit.confidence > prev.confidence) {
      best.set(hit.id, hit);
    }
  }
  const applied = [];
  for (const [id, hit] of best.entries()) {
    state.setCardBucket(id, hit.bucket);
    applied.push(hit);
  }
  const bottoms = parsed.accepted.filter((h) => h.location === "bottom" || h.location === "landlord_bottom");
  if (bottoms.length) {
    state.markBottom(bottoms.map((h) => h.id));
  }
  return {
    ...parsed,
    applied,
    message: `已合并 ${applied.length} 张识别牌面，跳过 ${parsed.skipped.length} 条`,
  };
}

export async function recognizeImage(_fileOrUrl) {
  throw new Error("图像识别模块尚未接入。请手动点选记牌，或导入识别结果文件。");
}

export function createVisionBridge(getState) {
  return {
    schemaVersion: VISION_SCHEMA_VERSION,
    schema: VISION_SCHEMA,
    recognizeImage,
    applyDetections(payload) {
      return ingestVisionPayload(getState(), payload);
    },
    parse: parseVisionPayload,
  };
}
